### PR TITLE
Moved dispatch to :gen_event

### PR DESCRIPTION
**Pull request by scohen**
As I was integrating indexing into lexical, I kept having to implement genservers that had no other point than to receive messages and take some action when one was received. :gen_event already does this, so I thought it would be nice to have a simple way to implement handler modules in addition to proceses that receive events.

### CODE REVIEW

File: apps/remote_control/lib/lexical/remote_control/dispatch/handler.ex [hunk 1]
```diff
@@ -0,0 +1,49 @@
+defmodule Lexical.RemoteControl.Dispatch.Handler do
```
Comment (zachallaun) [inline apps/remote_control/lib/lexical/remote_control/dispatch/handler.ex:1]: We should add a moduledoc here. It wasn't initially clear to me how this was intended to be used, until I looked at the tests. 

Suggestion:

```elixir
@moduledoc """
Defines a handler that selectively receives events emitted from a remote control node.

## Usage

Define a handler, specifying the events to be handled and implementing `on/2`:

    defmodule MyHandler do
      alias Lexical.RemoteControl.Dispatch.Handler
      alias Lexical.RemoteControl.Api.Messages
      
      import Messages

      use Handler, [project_compiled()]
      
      def on(project_compiled(), _state) do
        ...
        :ok
      end
    end

Register the handler with the dispatcher:

    Lexical.RemoteControl.Dispatcher.add_handler(MyHandler, :my_state)

"""
```

**Review by zachallaun — APPROVED** [review]
Made a couple suggestions but otherwise LGTM!

Just to confirm: we don't see any handler usage here (except for tests) because you're planning to use this in your upcoming indexing work, right?

File: apps/remote_control/lib/lexical/remote_control/dispatch/handler.ex (hunk 1 above)
Comment (zachallaun) [reply apps/remote_control/lib/lexical/remote_control/dispatch/handler.ex:1]: Also, we should make this a behaviour or change the moduledoc in Dispatch.

File: apps/remote_control/lib/lexical/remote_control/dispatch.ex [hunk 2]
```diff
@@ -1,137 +1,61 @@
 defmodule Lexical.RemoteControl.Dispatch do
-  defmodule State do
-    alias Lexical.Project
+  @moduledoc """
+  A global event dispatcher for lexical.
 
-    defstruct [:registrations]
+  Dispatch allows two recipients of its messages, processes and modules. A process must register
+  itself via a call to `register_listener`, while a process must implement the
+  `Lexical.RemoteControl.Dispatch.Handler` behaviour and add the module to the @handlers module attribute.
```
[the remark this answers is not in the document]
Comment (zachallaun) [reply apps/remote_control/lib/lexical/remote_control/dispatch.ex:7]: See comment on Handler: it's not currently a behaviour; we should make it one or change the doc here.

File: apps/remote_control/lib/lexical/remote_control/dispatch.ex [hunk 3]
```diff
@@ -1,137 +1,61 @@
...
-          {message, pids}
-        end)
+  # public API
 
-      %__MODULE__{state | registrations: registrations}
-    end
+  @doc """
+  Registers a process that will receive messages sent directly to its pid.
+  """
+  def register_listener(listener_pid, :all) do
```
[the remark this answers is not in the document]
Comment (scohen) [reply apps/remote_control/lib/lexical/remote_control/dispatch.ex:18]: eh, alright.

File: apps/remote_control/lib/lexical/remote_control/dispatch/handler.ex [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+          MapSet.member?(@event_names, :all) ->
+            on(event, state)
+
+          true ->
+            {:ok, state}
+        end
+      end
+
+      def handle_event(_event, state) do
```
[the remark this answers is not in the document]
Comment (scohen) [reply apps/remote_control/lib/lexical/remote_control/dispatch/handler.ex:41]: I kinda wanted to go ham on this and macro-up the `handle_event` callbacks, this is a remnant of that.

[the remark this answers is not in the document]
Comment (scohen) [reply apps/remote_control/lib/lexical/remote_control/dispatch/handler.ex:41]: Also, this won't crash the way you think under gen_event, the handler will be silently removed from the gen_event. I think logging an error message will be more effective.

**Pull request merged**